### PR TITLE
Added initial logic for Giant Swarm managed mode

### PR DESCRIPTION
**Pull request by AverageMarcus**
Signed-off-by: Marcus Noble <github@marcusnoble.co.uk>

This PR:

- adds/changes/removes etc

### Testing

Description on how aws-network-topology-operator can be tested.

- [ ] fresh install works
  - [ ] AWS
  - [ ] Azure
  - [ ] KVM
- [ ] upgrade from previous version works
  - [ ] AWS
  - [ ] Azure
  - [ ] KVM

#### Other testing

Description of features to additionally test for aws-network-topology-operator installations.

- [ ] check reconciliation of existing resources after upgrading
- [ ] X still works after upgrade
- [ ] Y is installed correctly

### Checklist

- [ ] Update changelog in CHANGELOG.md.
- [ ] Make sure `values.yaml` and `values.schema.json` are valid.

### CODE REVIEW

File: pkg/registrar/transitgateway.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+func (r *TransitGateway) Unregister(ctx context.Context, cluster *capi.Cluster) error {
+	logger := r.getLogger(ctx)
+
+	if annotations.IsNetworkTopologyModeGiantSwarmManaged(cluster) {
+		logger.Info("Network topology mode is GiantSwarmManaged")
+
+		// TODO: Check if TGW needs deleting
+
```
Comment (calvix) [inline pkg/registrar/transitgateway.go:145]: we will need to detach the VPC in case of WC 
if its MC than probably detach all VPCs and delete TGW

**Review by calvix — DISMISSED** [review]
LGTM

File: pkg/registrar/transitgateway.go (hunk 1 above)
Comment (AverageMarcus) [reply pkg/registrar/transitgateway.go:145]: Does a TGW need detaching before the VPC is deleted or will it just happen during delete?

Comment (AverageMarcus) [reply pkg/registrar/transitgateway.go:156]: Added anyway, can't hurt to detach first :)

File: pkg/registrar/transitgateway.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+			if tgw == nil {
+				err = fmt.Errorf("failed to find TransitGateway for provided ID")
+				logger.Error(err, "No TransitGateway found for ID provided on annotations", "transitGatewayID", gatewayID)
+				return err
+			}
+		}
+
+		// Ensure TGW ID is saved back to the current cluster
+		annotations.SetNetworkTopologyTransitGatewayID(cluster, *tgw.TransitGatewayId)
+		if cluster, err = r.clusterClient.Save(ctx, cluster); err != nil {
```
Comment (fiunchinho) [inline pkg/registrar/transitgateway.go:112]: same as above regarding patching

File: controllers/networktopology.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	Get(context.Context, types.NamespacedName) (*capi.Cluster, error)
+	AddFinalizer(context.Context, *capi.Cluster, string) error
+	RemoveFinalizer(context.Context, *capi.Cluster, string) error
+}
+
+//counterfeiter:generate . Registrar
+type Registrar interface {
+	Register(context.Context, *capi.Cluster) error
+	Unregister(context.Context, *capi.Cluster) error
+}
```
Comment (fiunchinho) [inline controllers/networktopology.go:31]: Why use the `Registrar` approach? will this operator create different resources?

**Review by fiunchinho — COMMENTED** [review]
Some comments and questions

File: controllers/networktopology.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		err = reg.Register(ctx, Cluster)
+		if err != nil {
+			if errors.Is(err, &registrar.ModeNotSupportedError{}) {
+				return ctrl.Result{Requeue: false}, nil
+			}
+			return ctrl.Result{}, microerror.Mask(err)
+		}
+	}
+
+	return ctrl.Result{Requeue: true, RequeueAfter: time.Minute * 10}, nil
```
[the remark this answers is not in the document]
Comment (AverageMarcus) [reply controllers/networktopology.go:94]: To ensure that outside changes to TGW are reflected. E.g. if this is a WC getting it's TGW from the MC we want to make sure updates to the MC TGW are respected.

File: pkg/registrar/transitgateway.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	gatewayID := annotations.GetNetworkTopologyTransitGatewayID(cluster)
+
+	switch val := annotations.GetAnnotation(cluster, annotations.NetworkTopologyModeAnnotation); val {
+	case "":
+		// If no value currently set, we'll set to the default of 'None"
+		logger.Info("NetworkTopologyMode is currently unset, setting to the default of 'None'")
+		annotations.AddAnnotations(cluster, map[string]string{
+			annotations.NetworkTopologyModeAnnotation: annotations.NetworkTopologyModeNone,
+		})
+		if _, err := r.clusterClient.Save(ctx, cluster); err != nil {
```
[the remark this answers is not in the document]
Comment (AverageMarcus) [reply pkg/registrar/transitgateway.go]: 👍  Done

File: pkg/registrar/transitgateway.go (hunk 2 above)
Comment (AverageMarcus) [reply pkg/registrar/transitgateway.go]: 👍  Done

File: controllers/networktopology.go (hunk 3 above)
Comment (AverageMarcus) [reply controllers/networktopology.go:31]: Possibly. I've been focussing on the GiantSwarmManaged mode for this which is pretty straight forward. When it comes to the UserManaged mode it's possible there's more we need to handle but I'm not sure at this stage.

Are there any strong reasons not to use it at this stage? Or do you think I should refactor it for now and add in when needed at a later date?

**Review by fiunchinho — APPROVED** [review]

**Pull request merged**